source/common/common/inline_map.h

```diff
@@ -411,44 +437,36 @@ template <class Key, class Value> class InlineMap : public InlineStorage {
     if (!inlineEntryValid(handle.inlineId())) {
       resetInlineMapEntry(handle.inlineId(), Value());
     }
-    return inline_entries_[handle.inlineId()];
-  }
-
-  /**
-   * Create an inline map with the given descriptor. If the descriptor is not finalized, it will be
-   * finalized before creating the inline map.
-   * @param descriptor the descriptor that contains the inline keys.
-   * @return the created inline map.
-   */
-  static std::unique_ptr<InlineMap> create(TypedInlineMapDescriptor& descriptor) {
-    if (!descriptor.finalized()) {
-      // Call finalize() to make sure that the descriptor is finalized and no any new inline
-      // keys could be registered.
-      descriptor.finalize();
-    }
-
-    return std::unique_ptr<InlineMap>(new ((descriptor.inlineKeysNum() * sizeof(Value)))
-                                          InlineMap(descriptor));
+    return inlineEntries()[handle.inlineId()];
   }
 
 private:
+  using StoragePtr = std::unique_ptr<uint8_t[]>;
   friend class InlineMapDescriptor<Key>;
 
-  InlineMap(TypedInlineMapDescriptor& descriptor) : descriptor_(descriptor) {
-    // Initialize the inline entries to nullptr.
-    uint64_t inline_keys_num = descriptor_.inlineKeysNum();
-    inline_entries_valid_.resize(inline_keys_num, false);
-    memset(inline_entries_storage_, 0, inline_keys_num * sizeof(Value));
-    inline_entries_ = reinterpret_cast<Value*>(inline_entries_storage_);
+  void lazyInitializeInlineEntriesMemory() {
+    if (inline_entries_ == nullptr) {
+      const uint64_t value_memory_size = descriptor_.inlineKeysNum() * sizeof(Value);
+      const uint64_t valid_memory_size = descriptor_.inlineKeysNum() * sizeof(bool);
+      const uint64_t memory_size = value_memory_size + valid_memory_size;
+
+      inline_entries_.reset(new uint8_t[memory_size]);
+      memset(inline_entries_.get(), 0, memory_size);
+
+      inline_entries_valid_ = reinterpret_cast<bool*>(inline_entries_.get() + value_memory_size);
```

Comment:
It's frequent operation to check the validity of inline entry. We keep this pointer to  avoid computing the offset every time.

Reply:
Again you are kind of halfway optimizing this. Note that arithmetic in the pipeline generally is quite fast compared to memory access.

But if you want to spend the memory to hold this extra pointer why don't you just use a separate allocation to avoid doing the kind of direct memory manipulation we usually try to avoid?

Reply:
I think the initial target is to avoid unnecessary memory allocation. Keeping the pointer is to avoid repeatedly offset computing.
If you think the offset computing is pretty fast, I can try to update it.

By the way, I just read the source code of flat hash map, I found almost same design.
A continuous memory is allocated for values and flags, and additional two typed pointers are used to store the first address of values and flags.

Reply:
🤔 because from current test results, the inline map was proved that it is not approprivate for the scenario where the instances will be created/destructed frequently. May be seperate allocations are acceptable, it won't make thing worse too much.

Although this way still cannot resolve the alignment problem.

Reply:
What alignment problem? I don't think you have one if you use malloc. The only exception might be if you have a data structure requiring SSE operations.